src/Microsoft.ServiceFabric.Services.Queryable/Controller/QueryableController.cs

```diff
@@ -79,28 +195,78 @@ private IHttpActionResult HandleException(Exception e, Uri serviceUri)
 				return BadRequest(e.Message);
 			if (e.InnerException is ArgumentException)
 				return BadRequest(e.InnerException.Message);
+			if (e is HttpException)
+				return Content((HttpStatusCode)((HttpException)e).GetHttpCode(), ((HttpException)e).Message);
+			if (e.InnerException is HttpException)
+				return Content((HttpStatusCode)((HttpException)e.InnerException).GetHttpCode(),
+					((HttpException)e.InnerException).Message);
 
 			if (e is AggregateException)
 				return InternalServerError(e.InnerException ?? e);
 
 			return InternalServerError(e);
 		}
 
+		private static readonly ThreadLocal<Random> random = new ThreadLocal<Random>(() => new Random());
+
 		private static async Task<T> GetServiceProxyAsync<T>(Uri serviceUri) where T : IService
 		{
 			using (var client = new FabricClient())
 			{
 				var partitions = await client.QueryManager.GetPartitionListAsync(serviceUri).ConfigureAwait(false);
-				return CreateServiceProxy<T>(serviceUri, partitions.First());
+				int randomindex = random.Value.Next(0, partitions.Count);
+				return CreateServiceProxy<T>(serviceUri, partitions[randomindex]);
+			}
+		}
+
+		private static async Task<IEnumerable<T>> GetServiceProxyForPartitionAsync<T>(Uri serviceUri, Guid partitionId)
+			where T : IService
+		{
+			using (var client = new FabricClient())
+			{
+				var partitions = await client.QueryManager.GetPartitionListAsync(serviceUri).ConfigureAwait(false);
+				var matchingPartitions =
+					partitions.Where(p => p.PartitionInformation.Id == partitionId || partitionId == Guid.Empty);
+				return matchingPartitions.Select(p => CreateServiceProxy<T>(serviceUri, p));
+			}
+		}
+
+		private static async Task<T> GetServiceProxyForDeleteAsync<T>(Uri serviceUri, Guid partitionId)
+			where T : IService
+		{
+			using (var client = new FabricClient())
+			{
+				var partitions = await client.QueryManager.GetPartitionListAsync(serviceUri).ConfigureAwait(false);
+				var matchingPartition = partitions.Where(p => p.PartitionInformation.Id == partitionId);
+				return CreateServiceProxy<T>(serviceUri, matchingPartition.First());
+			}
+		}
+
+		private static async Task<T> GetServiceProxyForAddAsync<T>(Uri serviceUri, Guid partitionId) where T : IService
+		{
+			using (var client = new FabricClient())
+			{
+				var partitions = await client.QueryManager.GetPartitionListAsync(serviceUri).ConfigureAwait(false);
+				var matchingPartitions = partitions.Where(p => p.PartitionInformation.Id == partitionId);
+
+				int randomindex = random.Value.Next(0, partitions.Count);
```

Comment:
If memory serves, the partition key is better if it's between INTMIN and INTMAX, though perhaps we don't actually have any guarantees about the distribution. @jessebenson Can you comment?